src/Features/Core/Portable/Completion/CommonCompletionService.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Snippets;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    internal abstract partial class CommonCompletionService : CompletionServiceWithProviders
+    {
+        protected CommonCompletionService(Workspace workspace)
+            : base(workspace)
+        {
+        }
+
+        // Magic trigger for VB snippets
+        internal static readonly CompletionTriggerKind SnippetsTriggerReason = (CompletionTriggerKind)(-1);
+
+        protected override ImmutableArray<CompletionProvider> GetProviders(ImmutableHashSet<string> roles, CompletionTrigger trigger)
+        {
+            var providers = base.GetProviders(roles, trigger);
+
+            if (this.Language == LanguageNames.VisualBasic)
+            {
+                if (trigger.Kind == SnippetsTriggerReason)
+                {
+                    providers = providers.Where(p => p.IsSnippetProvider).ToImmutableArray();
+                }
+                else
+                {
+                    providers = providers.Where(p => !p.IsSnippetProvider).ToImmutableArray();
+                }
+            }
+
+            return providers;
+        }
+
+        protected override CompletionItem GetBetterItem(CompletionItem item, CompletionItem existingItem)
+        {
+            // We've constructed the export order of completion providers so 
+            // that snippets are exported after everything else. That way,
+            // when we choose a single item per display text, snippet 
+            // glyphs appear by snippets. This breaks preselection of items
+            // whose display text is also a snippet (workitem 852578),
+            // the snippet item doesn't have its preselect bit set.
+            // We'll special case this by not preferring later items
+            // if they are snippets and the other candidate is preselected.
+            if (existingItem.Rules.Preselect && IsSnippetItem(item))
+            {
+                return existingItem;
+            }
+
+            // If one is a keyword, and the other is some other item that inserts the same text as the keyword,
+            // keep the keyword (VB only)
+            if (IsKeywordItem(existingItem) && this.Language == LanguageNames.VisualBasic)
```

Comment:
This doesn't seem like how we shoudl be doing htings.  It's been a core principle for how we've developed features that we don't put special cases for specific language cases at the common layer.  For all intents and purposes the common layer should have no idea what language it is working on.

--

This is also problematic because it's only possible for us to do.  Say TS or F# need this behavior, how would they get it?